avocado-framework/avocado
# [V10] Virt test compat layer

This is a follow up to PR #626

This is the most fleshed out virt test compat layer to date. It introduces loading of tests from multiple test sources, which allows the mix and match of regular avocado tests and avocado virt-tests:

	$ scripts/avocado run --vt-setup type_specific.io-github-autotest-qemu.migrate.default.tcp examples/tests/passtest.py --open-browser
	SETUP      : PASS (15.66 s)
	JOB ID     : 0a613e41201ee2fa726a6ef8617f6578aa901e3a
	JOB LOG    : /home/lmr/avocado/job-results/job-2015-04-30T01.09-0a613e4/job.log
	JOB HTML   : /home/lmr/avocado/job-results/job-2015-04-30T01.09-0a613e4/html/results.html
	TESTS      : 2
	(1/2) type_specific.io-github-autotest-qemu.migrate.default.tcp: PASS (89.62 s)
	(2/2) examples/tests/passtest.py: PASS (0.00 s)
	PASS       : 2
	ERROR      : 0
	FAIL       : 0
	SKIP       : 0
	WARN       : 0
	INTERRUPT  : 0
	TIME       : 89.62 s

You can also do cool things like listing all virt test `migrate` subtests:

        $ scripts/avocado list migrate --verbose
        Type Test                                                                  
        VT   type_specific.io-github-autotest-qemu.migrate.default.x-rdma                   
        VT   type_specific.io-github-autotest-qemu.migrate.default.rdma                     
        VT   type_specific.io-github-autotest-qemu.migrate.default.tcp                      
        VT   type_specific.io-github-autotest-qemu.migrate.default.unix                     
        VT   type_specific.io-github-autotest-qemu.migrate.default.exec.default_exec        
        VT   type_specific.io-github-autotest-qemu.migrate.default.exec.gzip_exec           
        VT   type_specific.io-github-autotest-qemu.migrate.default.fd                       
        VT   type_specific.io-github-autotest-qemu.migrate.default.mig_cancel               
        VT   type_specific.io-github-autotest-qemu.migrate.default.mig_cancel_x_rdma        
        VT   type_specific.io-github-autotest-qemu.migrate.default.mig_cancel_rdma          
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.x-rdma            
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.rdma              
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.tcp               
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.unix              
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.exec.default_exec 
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.exec.gzip_exec    
        VT   type_specific.io-github-autotest-qemu.migrate.with_set_speed.fd                
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.x-rdma               
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.rdma                 
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.tcp                  
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.unix                 
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.exec.default_exec    
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.exec.gzip_exec       
        VT   type_specific.io-github-autotest-qemu.migrate.with_reboot.fd                   

        ACCESS_DENIED: 0
        BROKEN_SYMLINK: 0
        BUGGY: 0
        INSTRUMENTED: 0
        MISSING: 0
        NOT_A_TEST: 0
        SIMPLE: 0
        VT: 24

You can try out this work by following the steps:

1) Check out this branch
2) Set the environment variable 'VIRT_TEST_PATH' to a path containing a copy of virt test
3) The copy of virt-test has to be in my new branch `avocado-integration-v3` [1]
4) Bootstrap your testing in the virt test dir by running `./run -t qemu --bootstrap`
5) Run your tests with the command above
6) Now you can use `scripts/avocado list` to see virt-tests listed there.

That's it. Virt Test at this point has way too many knobs, and it's challenging to look to each of them and make sure they are working well. I hope that at least you guys enjoy trying this out.
 
Known issues:
* Some unittests fail because now we say on stdout that plugins have failed to load.

Changes from v9:
* Now the plugin honors the old virt test setting of `$AUTOTEST_PATH`, in case you want to use autotest from a git repo instead of from an RPM. In order to get consistency with that option, we changed `$VIRT_TEST_DIR` -> `$VIRT_TEST_PATH`.
* `avocado list --vt-guests`: Filter by guest list `i440fx`, which prevents issues with the `q35` variant not listing tests
* Set priority of test lister to be 0, to prevent issue when the virt test lister is loaded before the test lister: `virt_test_compat_lister': 'Parser' object has no attribute 'lister'`

Changes from v8:
* Finally the `avocado list` functionality is complete. You can list tests by `scripts/avocado list`, use `--vt-type` (different test types generate different test lists) and even list guests com `scripts/avocado --vt-list-guests` (different guests generate different test lists).

Changes from v7:
* Allow avocado list to show tests from all test loader plugins. This way, when enabled, virt-tests are displayed. There's one last thing lacking here, how to add new arguments to the lister plugin. I know how to do that for the runner plugin, I'll need @ruda's help to understand how to do it.

Changes from v6:
* Fixed bug running virt tests with multiplex file specified in the command line
* Added an avocado_params attribute to the virt test class of the plugin, allowing users to use 'frankenstein mode' - access avocado specific muliplexed params from inside virt-tests.
* Removed the aggressive error message when the virt plugin is not configured.

Changes from v5:
* Cleaned up the plugin code and adapted it to the refactorings I made on `virt-test`'s side.

Changes from v4:
 * Finally resolved the problem with loading multiple test sources, while keeping the base filesystem plugin working the way it usually does. With this, unittest hacks could be thrown away. Unfortunately, the overhead with the virt plugin enabled makes a couple of unittests to fail, but since the compat plugin is going to be on a separate project, that isn't a big problem per se. Now, I'm certain that there are ways to make the multiple test loading more efficient. Figuring it all out at once is not easy, though.

Changes from v3:
 * Removed the list and bootstrap plugins. virt-test can already do that functionality for the user.
 * Went to a excruciatingly detailed review of command line options. Many options that don't make sense in avocado were removed, text reviewed, and options grouped logically. Next step is to move unneeded options to the config file.

[1] See https://github.com/autotest/virt-test/pull/2176

Bonus: People can now generate HTML reports out of virt-test runs. I imagine there will be people happy with this.

![screenshot from 2015-04-30 01-22-14](https://cloud.githubusercontent.com/assets/296807/7406339/7699689e-eed7-11e4-9214-38a678c105ec.png)
